src/Promise.php

```diff
@@ -27,15 +33,45 @@ public function then(callable $onFulfilled = null, callable $onRejected = null):
             return new static($this->resolver($onFulfilled, $onRejected));
         }
 
-        $this->requiredCancelRequests++;
+        // This promise has a canceller, so we create a new child promise which
+        // has a canceller that invokes the parent canceller if all other
+        // followers are also cancelled. We keep a reference to this promise
+        // instance for the static canceller function and clear this to avoid
+        // keeping a cyclic reference between parent and follower.
+        $parent = $this;
+        ++$parent->requiredCancelRequests;
+
+        return new static(
+            $this->resolver($onFulfilled, $onRejected),
+            self::parentCancellerFunction($parent)
```

Comment:
It looks like some changes have not been cherry-picked here? https://github.com/reactphp/promise/commit/9281a8a03b71d905a72adb8f4bea2e7abb87bf80

Reply:
See also https://github.com/reactphp/promise/commit/72087fdb3686cf99b4404b26074f6a2aa0ff11b1#diff-71f7b9a6df57aaf364b33eacad82e245R64

Reply:
Are those the only two?

Reply:
Performed a quick diff between your branch and the `2.x` branch and these where the only ones that caught my attention, perhaps double check this to confirm?

Reply:
Ok cool, yes of course. The thing is certain other things also changed to need to work in a way where this all still works out in the end :)

Reply:
@clue ok should be fixed now, can you have another look?